### PR TITLE
refactor: Switched to WMR, largely following Preact style now

### PR DESCRIPTION
**Pull request by rschristian**
Still have a lot of items to touch up, but getting there.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,123 +1,36 @@
...
-  "license": "MIT",
-  "homepage": "https://github.com/mutoe/preact-realworld-example-app#readme",
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/mutoe/preact-realworld-example-app.git"
-  },
-  "bugs": {
-    "url": "https://github.com/mutoe/preact-realworld-example-app/issues"
-  },
-  "main": "src/index.tsx",
```
Comment (rschristian) [inline package.json:20]: Just so you know for the future, all of these are only relevant for packages. They have 0 reason to exist on an application. It's meta data for package repositories like https://npmjs.com

File: .github/workflows/nodejs.yml [hunk 2]
```diff
@@ -1,57 +0,0 @@
-name: Node CI
```
Comment (mutoe) [inline .github/workflows/nodejs.yml]: I don't think the CI setup needs to remove. We should provide a case on how to set up a pipeline example, right?

**Comment by mutoe** [comment]
@rschristian  So, what do you think is the purpose of the Realworld code repository? For others to learn how to work with Preact or to demonstrate the smallest usable example?

File: .github/workflows/nodejs.yml (hunk 2 above)
Comment (rschristian) [reply .github/workflows/nodejs.yml]: How is a pipeline really relevant to Preact? I would say no, we shouldn't, though I'll add it back regardless.

**Comment by rschristian** [comment]
Realworld is a way to show off a language/framework/lib in an actual setting, not some contrived example (which tends to come from the official docs). This means to demonstrate how the language/tool is unique and offers different options and advantages with a quick glance.

The problem that I'm looking to solve here is that what currently exists doesn't show off any of Preact's advantages, and instead focuses on generic items that are unrelated. Just because items like ESLint or CI pipelines exist and arguably have some value does not mean they need to be displayed here.

For example, what about the ESLint config or the CI pipeline you've set up differs from literally any other Node application? What makes it Preact-centric? You could copy those to any other Node project and they'd be just as valid. Spending time on those items takes away from what users are really here to see: How does this differ? How is this unique?

Furthermore, many of Preact's advantages are squandered here. 

 - Preact's JSX can be turned into `h()` with some light babel magic which vastly improves upon build speed, not to mention gives the possibility of a truly great HMR experience
   - Only the repo uses `ts-loader` which just treats it like standard JSX so that advantage is null and void. Not to mention no HMR.
 - Preact's build size is really what sets it apart as it punches way above it's weight
   - `history` is used which eats up that entire advantage as it's quite a poor package.
 - Preact is much closer to the DOM, giving users the option and ability to treat their JSX as if it was HTML. No more need for renamed attributes or events.
   - `className` is used everywhere as is the React-specific naming 

None of the Preact code style has been adopted in or any of the build tools custom made to make Preact a great experience.

I'm not trying to rag on you or say this is poor or anything, this was (I imagine) a ton of work to set up and introduce when you could've easily just not. But we should going forward look to make this a repo that shows off Preact, and what it can do, rather than "here's a generic Node CI system that runs `yarn test` as it would in any Node project". 

That's all I'm looking to do here. Improve the experience for someone who clicks on this having never heard of Preact before. Let's get them into the meat of what makes it unique rather than bogging them down in generic boilerplate.

**Comment by rschristian** [comment]
Might need to sit on this for a bit. Preact-ISO (like a toolkit built for WMR, has routing, prerendering, hydration, etc.) is getting a few updates shortly here that'll we'll need/should make use of, so will wait for a release on those.

Jest is also a mess currently, not really supporting ESM. Might end up having to transpile `node_modules` unfortunately as a few deps ship with `"type": "module"` and Jest refuses to acknowledge that they're modules.

File: public/pages/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import { useEffect, useState } from 'preact/hooks';
+
+import { useRootState } from '../store';
+import { login } from '../store/actions';
+
+export default function Register() {
+    const formRef = createRef<HTMLFormElement>();
```
Comment (rschristian) [inline public/pages/Login.tsx:7]: This is quite bad, if you did not know. 

Only use `createRef()` in class based components. `createRef()` will always create a new reference, and this is fine in class-based components where you have a constructor. In functional components though, this means you are creating a new reference on every single render. This is where `useRef()` comes in. It'll return the same ref upon every render.

The names are super similar so might've just been a typo in the suggestion from your IDE for all I know. Just thought I'd mention it.

Comment (mutoe) [reply public/pages/Login.tsx:7]: You are right. It's a mistake.

**Comment by mutoe** [comment]
Maybe you are right. I noticed that you are from the Preact team, and you should adjust this project in accordance with the official recommendations. I will give you permission to cooperate.

What I want to say is that as a user, I just spent a long time setting up the development environment, so I want Realworld to become a reference configuration project, not a minimal usable example. Maybe I was wrong, just for reference~

**Comment by rschristian** [comment]
I'm not looking to dictate or demand or anything, just to make that clear in case any of the comments came off too forward. I sincerely apologize if they have, that's not what I've intended.

You've definitely changed my mind on a few things and discussion is good. It increases quality and that's all I'm ultimately looking to do, improve quality/user experience/onboarding or whatever if I can. Really I'm only here to make suggestions.

**Comment by mutoe** [comment]
The WMR you proposed is a good idea. I am currently working on a Vue3 project, and I even want to use Vite to replace Webpack in this project. It doesn't seem to be necessary now.

Go ahead and do it!

**Comment by rschristian** [comment]
Ah, yeah Vite is super capable too, used it a bit. And, if you didn't know, its [plugin system actually came from WMR](https://github.com/vitejs/vite/blob/5ec13d8b3fe1632f793c7ad22b21c43a13d71141/packages/vite/src/node/server/pluginContainer.ts#L1-L11), so obviously the projects share many similarities. 

WMR has a few advantages for Preact, given it comes from the Preact team, but otherwise definitely pretty similar.

**Comment by mutoe** [comment]
Cool! It turns out that the Preact team is so good! 👍

**Comment by rschristian** [comment]
Have a decent idea of the plan now, will be trickling it in. No need for this PR.

**Pull request closed without merging**